### PR TITLE
Lab1 - Fibonacci function

### LINKED ISSUE #193: Lab 1 Fibonacci sequence
**Lab 1 - Fibonacci sequence**
- [ ] Create an executable go program in directory 01_fib
- [ ] Write a function that prints the first n Fibonacci numbers
```
func fib(n int)
```
- [ ] Call fib(7) in main to print
```
1
1
2
3
5
8
13
```
- [ ] Bonus points: For negative n print Negafibonacci numbers.

### PR DESCRIPTION
**Pull request by Jake88**
Note: I understand storing a map of all the calculated fibonacci numbers is inefficient given the criteria is only to _print_ them, so I have over-engineered here.

Fixes #193

Review of colleague's PR #183

#### Changes proposed in this PR:
- Solution to lab1 - Fibonacci sequence
- Tests for the fib function

### CODE REVIEW

File: 01_fib/Jake88/main.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+)
+
+var out io.Writer = os.Stdout
+
+func main() {
+	fib(7)
+}
+
+func fib(n int) {
+	// Our intial map, with the default starting values needed to begin the sequence.
```
Comment (golangcibot) [inline 01_fib/Jake88/main.go]: `intial` is a misspelling of `initial` (from `misspell`)

**Comment by willshen8** [comment]
Hi Jake, can I do the review your PR once it's passed CI? For now, can you please remove 'Ready for review' label?

**Comment by Jake88** [comment]
> Hi Jake, can I do the review your PR once it's passed CI? For now, can you please remove 'Ready for review' label?

More than happy for you to review my changes! :) I've fixed up my typo now, so go for gold.

I do think I need to rebase (or something) my typo commit, but I honestly don't know how to go about doing that so I might need to ask an adult for help!

**Comment by willshen8** [comment]
Hi Jake,
All looks good. Few things to consider:

1. Map would occupy more computing resources compare to array or slice, so computing a larger fib number might have efficiency differences?
2. If you compute large fib number, you might get integer overflow from int, so change it to a larger native type. 
3. A good programming practices is seperation of concerns. You could have a function that generate the map and another function does the print. So that these two functions can be resued if the program gets larger.

**Comment by Jake88** [comment]
> Hi Jake,
> All looks good. Few things to consider:
> 
> 1. Map would occupy more computing resources compare to array or slice, so computing a larger fib number might have efficiency differences?
> 2. If you compute large fib number, you might get integer overflow from int, so change it to a larger native type.
> 3. A good programming practices is seperation of concerns. You could have a function that generate the map and another function does the print. So that these two functions can be resued if the program gets larger.

Thanks!
1. You're right here. In hindsight I should have stuck to the criteria and just stored the last two numbers in the sequence, since we don't need to retain or return the full list!
2. I didn't really think about error handling or scaling too high! I should have tried to break it first and then I would have seen these gaps. ;) Good spot.
3. Sure. I think my fib function is definitely more complex than it needs to be. I guess if the program needed to scale I would look at taking reusable pieces and putting them into their own function.

**Comment by anzboi** [comment]
Please refrain from merging master into your branch until you have the `merge me` label added by an instructor. This way you will have less master merges in your PR.

File: 01_fib/Jake88/main.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+func fib(n int) {
+	// Our initial map, with the default starting values needed to begin the sequence.
+	fibonacciMap := map[int]int{
+		-1: 1,
+		0:  0,
+	}
+	negative := false
+
+	// Check if we are performaing negafibonacci or not, and if so remove the sign from n
```
Comment (anzboi) [inline 01_fib/Jake88/main.go]: s/performaing/performing

File: 01_fib/Jake88/main.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+var out io.Writer = os.Stdout
+
+func main() {
+	fib(7)
+}
+
+func fib(n int) {
+	// Our initial map, with the default starting values needed to begin the sequence.
+	fibonacciMap := map[int]int{
```
Comment (anzboi) [inline 01_fib/Jake88/main.go]: This map is overkill. You only use two of its value at any given time, and it is never used once the function returns.
I would suggest you store two variables (perhaps `current` and `previous`) initialized to `1` and `0`, then update them in the loop accordingly.
Alternatively, you can store the map as an unexported `var` outside this function to turn it into a global variable,
```
var fibonacciMap = map[int]int{
  -1: 1,
  0: 0,
}
```
If you go down this path, you will find it harder to write the fibonacci function (and may find it easier to write a recursive function that first checks the map, then populates it if it doesn't exist before returning).
I wouldn't recommend this though.

File: 01_fib/Jake88/main.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+		// Store the next positive fibonacci iteration
+		current := fibonacciMap[i-1] + fibonacciMap[i-2]
+		fibonacciMap[i] = current
+
+		// If negafibonacci, store the next negative iteration
+		if negative {
+			if i%2 == 0 {
+				current = -current
+			}
+			fibonacciMap[-i] = current
```
Comment (anzboi) [inline 01_fib/Jake88/main.go]: The map values are never accessed for negative integers (except `-1` on the first iteration). This is storing numbers that will never be read.

File: 01_fib/Jake88/main_test.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+	type test struct {
+		input    int
+		expected string
+	}
+
+	tests := []test{
+		{-7, `"1\n-1\n2\n-3\n5\n-8\n13\n"`},
+		{7, `"1\n1\n2\n3\n5\n8\n13\n"`},
+		{-3, `"1\n-1\n2\n"`},
+		{4, `"1\n1\n2\n3\n"`},
```
Comment (anzboi) [inline 01_fib/Jake88/main_test.go:37]: Can we add a test for `fib(0)`.
I also don't think you need the backticks here (correct me if I am wrong).

**Review by anzboi — CHANGES_REQUESTED** [review]
Great attempt so far, love that you are playing around with various objects in `go` 👍 .
For the purposes of the lab though, I would like you to rethink the use of the map inside the `fib` function. See my comments for more details.
Also please test any edge cases in your labs (`fib(0)` in this case).

File: 01_fib/Jake88/main.go (hunk 3 above)
Comment (Jake88) [reply 01_fib/Jake88/main.go]: Actioned. Converted to two variables instead.

File: 01_fib/Jake88/main_test.go (hunk 5 above)
Comment (Jake88) [reply 01_fib/Jake88/main_test.go:37]: Actioned. Added both a test and functionality to handle fib(0).

File: 01_fib/Jake88/main.go (hunk 2 above)
Comment (Jake88) [reply 01_fib/Jake88/main.go]: Actioned. Proof read my new comments and made fixes accordingly.

File: 01_fib/Jake88/main.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+var out io.Writer = os.Stdout
+
+func main() {
+	fib(7)
+}
+
+func absolute(n int) (int, bool) {
+	var negative bool
+	if n < 0 {
+		n *= -1
```
Comment (anzboi) [inline 01_fib/Jake88/main.go:18]: `n=-n` works just as well.

File: 01_fib/Jake88/main.go [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	prev2, prev1 := 1, 0
+
+	n, nega := absolute(n)
+
+	for i := 0; i < n; i++ {
+		current := prev1 + prev2
+
+		// Protect against integer overflow / out of bounds error.
```
Comment (anzboi) [inline 01_fib/Jake88/main.go:38]: 👍

File: 01_fib/Jake88/main_test.go [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+
+		if test.expected != actual {
+			t.Errorf("Unexpected output printed by fib.\n(Expected) : %v\n(Actual)   : %v", test.expected, actual)
+		}
+
+		buf.Reset()
+	}
+
+	// Test out of bounds
+	fib(100)
```
Comment (anzboi) [inline 01_fib/Jake88/main_test.go:53]: 👍 for testing implemented functionality even if not required.
Normally we would write a separate test for this though.

**Review by anzboi — APPROVED** [review]
LGTM. Just one thing to think about for future labs.
When testing different behaviour (out of bounds vs normal), it should go in a different test.
